crowbar_framework/spec/lib/crowbar/lock/shared_non_blocking_spec.rb

```diff
@@ -0,0 +1,218 @@
+    # end
+    klass.new(defaults.merge(options))
+  end
+
+  after(:each) do
+    lock.release
+    lock.clean
+  end
+
+  context "when a lock is acquired" do
+    it "returns a lock object" do
+      expect(lock.acquire).to be_an_instance_of(klass)
+    end
+
+    it "sets a lock object to locked" do
+      expect(lock).to be_unlocked
+      lock.acquire
+      expect(lock).to be_locked_by_me
+    end
+
+    it "works via a #with_lock block" do
+      expect(lock).to be_unlocked
+      lock.with_lock do
+        expect(lock).to be_locked_by_me
+      end
+      expect(lock).to be_unlocked
+    end
+
+    it "is a shared lock" do
+      lock2 = new_lock(path: lock.path, owner: lock.owner + "2")
+      lock.acquire
+      lock2.acquire
+      expect(lock).to be_locked_by_me
+      expect(lock2).to be_locked_by_me
+      lock2.release
+      expect(lock).to be_locked_by_me
+      expect(lock2).to be_locked_by_another
+    end
+
+    it "keeps locks with different paths independent" do
+      # Check that one of two attempts racing to obtain lock will win
+      lock2 = new_lock
+      lock.acquire
+      lock2.acquire
+      expect(lock2).to be_locked_by_me
+      lock2.release
+      lock2.clean
+    end
+
+    it "protects against race conditions" do
+      lock1 = new_lock(executor: racy_local_executor(2, 0))
+      lock2 = new_lock(
+        executor: racy_local_executor(0.5, 0),
+        owner: lock1.owner + "2",
+        path: lock1.path
+      )
+
+      thread = Thread.new { lock1.acquire }
+      sleep 0.5 # make sure lock1 gets the flock first
+      lock2.acquire # this should wait for lock1 to be acquired
+      expect(lock1).to be_locked_by_me
+      expect(lock2).to be_locked_by_me
+      thread.join # this should happen instantly
+
+      lock1.release
+      lock2.release
+      lock1.clean
+    end
+  end
+
+  context "when a lockfile is released" do
+    it "returns a lock object" do
+      lock.acquire
+      expect(lock.release).to be_an_instance_of(klass)
+    end
+
+    it "unlocks a lock object" do
+      lock.acquire
+      lock.release
+      expect(lock).to be_unlocked
+    end
+
+    it "protects against race conditions" do
+      lock1 = new_lock(executor: racy_local_executor(2, 0))
+      lock2 = new_lock(
+        executor: racy_local_executor(0.5, 0),
+        owner: lock1.owner + "2",
+        path: lock1.path
+      )
+
+      lock1.acquire
+      lock2.acquire
+
+      thread = Thread.new do
+        lock1.release
+        expect(lock1).to be_locked_by_another
+      end
+      sleep 0.5 # make sure lock1 gets the flock first
+      lock2.release
+      expect(lock1).to be_unlocked
+      expect(lock2).to be_unlocked
+      thread.join
+
+      lock1.clean
+    end
+  end
+
+  context "when there is a stale lock from the same owner" do
+    before(:each) do
+      Dir.mkdir lock.readers_dir unless File.directory? lock.readers_dir
+      File.open(File.join(lock.readers_dir, lock.owner), 'w') do |f|
```

Comment:
Style/StringLiterals: Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.